src/metabase/query_processor/middleware/visualization_settings.clj

```diff
@@ -0,0 +1,72 @@
+(defn- update-card-viz-settings
+  "For each field, fetch its settings from the QP store, convert the settings into the normalized form
+  for visualization settings, and then merge in the card-level column settings."
+  [fields column-viz-settings]
+  (map (fn [[_ id _]]
+         (->> id
+              qp.store/field
+              :settings
+              (#(normalize-field-settings id %))
+              (#(m/deep-merge % (select-keys column-viz-settings [{::mb.viz/field-id id}])))))
+   fields))
+
+(defn- update-native-query-viz-settings
+  "For each column, select its viz settings from the column-viz-settings map by its name."
+  [columns column-viz-settings]
+  (map (fn [column] (-> column
+                        :name
+                        (#(select-keys column-viz-settings [{::mb.viz/column-name %}]))))
+       columns))
+
+(defn- get-viz-settings
+  "Pull viz settings from either the query map or the DB"
+  [query]
+  (if-let [card-id (-> query :info :card-id)]
+    ;; Saved card -> fetch viz settings from DB
+    (mb.viz/db->norm (db/select-one-field :visualization_settings Card :id card-id))
+    ;; Unsaved card -> viz settings passed in query
+    (-> query :viz-settings)))
+
+(defn- is-export-context?
+  [context]
+  (contains? #{:json-download :csv-download :xlsx-download} context))
+
+(defn update-viz-settings
+  "Middleware for fetching, processing and ordering a table's visualization settings so that they can be incorporated
+  into an export.
+
+  Card-level visualization settings are either fetched from the DB (for saved cards) or passed from the frontend
+  in the API call (for unsaved cards). These are merged with the base viz settings for each field that are fetched from
+  the QP store (and defined in the data model settings), and are ordered based on the field list from the query.
+
+  For native queries, only viz settings passed from the frontend are incorporated.
+  TODO: native queries only pass viz settings if columns have been hidden or reordered. This needs to be fixed on the frontend."
+  [qp]
+  (fn [query rff context]
+    (if (is-export-context? (-> query :info :context))
+      (let [viz-settings        (get-viz-settings query)
+            column-viz-settings (::mb.viz/column-settings viz-settings)]
+        (cond
+          (= (:type query) :query)
+          (let [fields (-> query :query :fields)
+                ordered-col-viz-settings (update-card-viz-settings fields column-viz-settings)
+                rff' (fn [metadata] (rff (assoc metadata :ordered-col-viz-settings ordered-col-viz-settings)))]
+            (qp query rff' context))
+
+          (= (:type query) :native)
+          (let [columns                  (-> viz-settings :table.columns)
+                ordered-col-viz-settings (update-native-query-viz-settings columns column-viz-settings)
+                rff' (fn [metadata] (rff (assoc metadata :ordered-col-viz-settings ordered-col-viz-settings)))]
+            (qp query rff' context))))
```

Comment:
The `rff'` and `(qp query rff' context)` code is the same between the two branches here... it looks like only `fields`/`column` is different. I'd push the `cond`/condp =` down to that point so you don't need to duplicate all the other code.

Reply:
The function called to generate `ordered-col-viz-settings` is also different between the branches. But I think I see what you mean.